### PR TITLE
Flash Attention v2

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
## Note
This has brought the FA2 kernel up until e07aa036db0e6cd251b17d9ee2d1720023da2def which has changed the semantics for causal masking when seq_q is not equal to seq_k
https://github.com/Dao-AILab/flash-attention/pull/436

## PR Dependencies 
I don't use ghstack :( this is a PR where it would have been helpful. That beings said I am going to peel off some PRs to make reviewing this easier:
- [x] Separate build flags for Flash and MemEff: #107985

### Description
This pull request updates the version of _scaled_dot_product_flash_attention from version 1 to version 2. The changes are based on the flash attention code originally authored by @tridao 

### Changes Made
The majority of the changes in this pull request involve:

- Copying over the flash_attention sources.
- Updating header files.
- Removing padding and slicing code from within the flash_attention kernel and relocating it to the composite implicit region of the SDPA. This was need to make the kernel functional and appease autograd.
- Introducing a simple kernel generator to generate different instantiations of the forward and backward flash templates.
- Adding conditional compilation (ifdef) to prevent building when nvcc is invoked with gencode < sm80.
- Introducing a separate dependent option for mem_eff_attention, as flash_attention v2 lacks support for Windows and cannot be built for sm50 generation codes.
- Modifying build.sh to reduce parallelization on sm86 runners and to lower the maximum parallelization on the manywheel builds. This adjustment was made to address out-of-memory issues during the compilation of FlashAttentionV2 sources.
- Adding/Updating tests.

### Notes for Reviewers
This is not a fun review, and I apologize in advance.
Most of the files-changed are in the flash_attn/ folder. The only files of interest here IMO:
- aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp
- aten/src/ATen/native/transformers/cuda/flash_attn/kernels/generate_kernels.py ( this has been incorporated upstream to flash-attention github)

There are a number of files all related to avoiding OOMs in CI/CD. These are typically shell scripts.

### Follow up items
- Include the updates from https://github.com/Dao-AILab/flash-attention/commit/e07aa036db0e6cd251b17d9ee2d1720023da2def and https://github.com/Dao-AILab/flash-attention/commit/9e5e8bc91e30af5cdc321362b553f6c0da332e30 | https://github.com/pytorch/pytorch/issues/108108

### Work Items
- [x] I don't think Windows will be supported for 3.1.0 - Need to update cmakee
- [x] Let multi_query/attention pass through and test | UPDATE: I have the fast path implemented here: https://github.com/pytorch/pytorch/pull/106730 but since this will require changes to semantics of math to call repeat_interleave, I think this should be done as a followup.
- [x] Had to drop cutlass back to 3.0.0 to get it to compile. Need to figure out how to upgrade to 3.1.0 and later. Spoke with Tri and he is going to be taking a look. Note: compiling with clang currently errors for the cute headers. 
- [x] Update test exercise above codepath
- [x] Still need to disable on seq_len % 128 != 0 for backward( Tri beat me to it https://github.com/Dao-AILab/flash-attention/commit/a4f148b6abe077cd387e69e5caae45d63a76bb24)
- [x] Add determinism warning to BWD, Tri got to this one as well: 1c41d2b
- [x] Update dispatcher to universally prefer FlashV2
- [x] Update tests to exercise new head_dims
- [x] Move the head_dim padding from kernel to top level composite implicit function in order to make it purely functional
- [x] Create template generator script
- [x] Initial cmake support for building kernels/ folder
- [x] Replay CudaGraph changes

#### Forward only
The TFlops are reported here are on a100 that is underclocked.
flashv2_tflops_vs_seq_len

#### Forward+Backward
Ran a sweep and for large compute bound sizes we do see a ~2x performance increase for forw+back.
Screenshot 2023-07-20 at 3 47 47 PM

cc @ezyang @gchanan @ptrblck @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -538,6 +540,27 @@ inline void validate_sdpa_input(
   }
   return;
 }
+
+// FlashAttentionV2 requires that head dimension be a multiple of 8
+// This was previously done within the kernel, however
+// This causes the kernel to maybe alias query, key, value
```
Comment (ekagra-ranjan) [inline aten/src/ATen/native/transformers/attention.cpp:546]: What does "alias query, key, value" mean? And what is "composite region" in the next comment?

**Comment by ekagra-ranjan** [comment]
@drisspg Have you tried benchmarking just the forward pass? Can you share if we still get the ~2x improvement?

**Comment by drisspg** [comment]
SShed into runner and using docker stats, indeed we OOm when compiling the flash_attention kernels causing the docker container to crash: Probs need something similar to: https://github.com/Dao-AILab/flash-attention/commit/9c531bdc0a3404369ead182af6f8558b3e860162

**Comment by drisspg** [comment]
Ussing massif locally I am getting

``` Shell
--------------------------------------------------------------------------------
Command:            python setup.py develop
Massif arguments:   --time-unit=B
ms_print arguments: massif.out.1963663
--------------------------------------------------------------------------------


    MB
37.82^                                                                       #
     |                                                                       #
     |                                                                       #
     |                                                                       #
     |                                                                       #
     |                                                                      @#
     |                                                                      @#
     |                                                                      @#
     |                                                                      @#
     |                                                                      @#
     |                                                                      @#
     |                                                                      @#
     |                                                                      @#
     |                            @                                        @@#
     |             :::@@::::::@:::@:::::::::::::@@:::::::::@:::::::::::::::@@#
     | :@@@@:::::::: :@ :: :: @:: @:: : : : :: :@ ::: ::: :@:: :::: :::: ::@@#
     | :@ @ ::: :: : :@ :: :: @:: @:: : : : :: :@ ::: ::: :@:: :::: :::: ::@@#
     | :@ @ ::: :: : :@ :: :: @:: @:: : : : :: :@ ::: ::: :@:: :::: :::: ::@@#
     | :@ @ ::: :: : :@ :: :: @:: @:: : : : :: :@ ::: ::: :@:: :::: :::: ::@@#
     | :@ @ ::: :: : :@ :: :: @:: @:: : : : :: :@ ::: ::: :@:: :::: :::: ::@@#
   0 +----------------------------------------------------------------------->GB
     0                                                                   5.321

Number of snapshots: 58
 Detailed snapshots: [2, 3, 11, 16, 19, 28, 36, 49, 50, 51, 52, 53, 54, 55, 56 (peak)]

```

I don't know why this OOMS

File: CMakeLists.txt [hunk 2]
```diff
@@ -727,6 +727,11 @@ cmake_dependent_option(
   "Whether to build the flash_attention kernel for scaled dot product attention" ON
   "USE_CUDA AND NOT ROCM AND NOT CMAKE_CUDA_COMPILER_VERSION VERSION_LESS 11.6" OFF)
 
+if(USE_FLASH_ATTENTION)
+  message("USE_FLASH_ATTENTION SET, setting max jobs to 1")
+  set(ENV{MAX_JOBS} "1")
```
Comment (huydhn) [inline CMakeLists.txt:732]: This approach looks right to me if the OOM issue is caused by cmake uses too many parallel processes and doesn't limit itself to the amount of available memory.  Is there a way to set this to MAX_JOBS / 2 or MAX_JOBS - 4 here, same as what we are using to build libtorch https://github.com/pytorch/pytorch/blob/main/.ci/pytorch/build.sh#L301-L305?

File: .ci/pytorch/build.sh [hunk 3]
```diff
@@ -28,6 +28,10 @@ echo "Environment variables:"
 env
 
 if [[ "$BUILD_ENVIRONMENT" == *cuda* ]]; then
+ # Flash Attention if OOming lets try reducing the max jobs
+  # MAX_JOBS=$(nproc --ignore=4)
+  MAX_JOBS=1
```
Comment (drisspg) [inline .ci/pytorch/build.sh:33]: remove this

Comment (bhack) [reply .ci/pytorch/build.sh]: Is there any connection to https://circleci.canny.io/cloud-feature-requests/p/have-nproc-accurately-reporter-number-of-cpus-available-to-container ?

Comment (drisspg) [reply .ci/pytorch/build.sh]: somewhat related but the number of cpus available to the container would still be too high to avoid the OOM

File: .ci/pytorch/build.sh [hunk 4]
```diff
@@ -159,6 +159,12 @@ if [[ "$BUILD_ENVIRONMENT" == *cuda* && -z "$TORCH_CUDA_ARCH_LIST" ]]; then
   exit 1
 fi
 
+# We only build FlashAttention files for CUDA 8.0, and they require large amounts of
+# memory to build and will OOM
+if [[ "$BUILD_ENVIRONMENT" == *cuda* ]] && [[ "$TORCH_CUDA_ARCH_LIST" == *"8.6"* ]]; then
+  export MAX_JOBS=2
```
Comment (drisspg) [inline .ci/pytorch/build.sh:165]: Try settings this to 3 again

File: .ci/pytorch/build.sh (hunk 3 above)
Comment (drisspg) [reply .ci/pytorch/build.sh]: you are correct, thats why in the updated code I just set the max_jobs manually based off of known amount of the container memory

Comment (bhack) [reply .ci/pytorch/build.sh]: https://discuss.circleci.com/t/can-you-if-on-resource-class/40581

File: test/inductor/test_fused_attention.py [hunk 5]
```diff
@@ -250,6 +251,7 @@ def sfdp_pattern_8(query, key, value):
         self._check_common(sfdp_pattern_8, args, atol=2e-3)
 
     @skipIfRocm
+    @unittest.skip("FlashV2 fails since the minimal mask block is 32")
```
Comment (drisspg) [inline test/inductor/test_fused_attention.py:254]: cc @eellison

Comment (drisspg) [reply test/inductor/test_fused_attention.py:254]: Need to fact check what I said in this comment. I am sure though that it fails with 32 elements because of dropout differences but gotta determine why.

Still don't think the test is designed properly but need to root cause.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 6]
```diff
@@ -725,10 +720,11 @@ _scaled_dot_product_flash_attention_nestedtensor_cuda(
           return_debug_mask,
           scale);
   // Reshape output to convert nnz to batch_size and seq_len
+  // Don't reshape {q,k,v}_padded since they are only used for backwards
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:723]: Aside from this one comment, all the other changes seem cosmetic?

File: aten/src/ATen/native/transformers/attention.cpp [hunk 7]
```diff
@@ -629,6 +630,37 @@ at::Tensor preprocess_mask(
 
   return attn_mask;
 }
+// FlashAttentionV2 requires that head dimension be a multiple of 8
+// This was previously done within the kernel, however
+// This causes the kernel to maybe alias query, key, value
+// So instead we pad the head_dimensions to be a multiple of 8 in the composite
+// region
+template <int alignment_size, bool slice>
+at::Tensor pad_last_dim(const at::Tensor& attn_bias) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/attention.cpp:639]: Why is the head dimension the last dim? I though that's the embedding dimension.

Comment (cpuhrsch) [reply aten/src/ATen/native/transformers/attention.cpp:639]: I'm probably just mixing up terminologies here.

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp (hunk 6 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:723]: you are correct, in fact this is an old comment. And I tweaked the behavior of the flash_attention slightly in order to make it conform to the autograd invariants. I left a comment in the summary but essentially all the padding happens in the top level sdpa where it will get automatically tracked by autograd.

File: aten/src/ATen/native/transformers/attention.cpp (hunk 7 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/attention.cpp:639]: so by the time you have called sdpa, you have taken your `embed_dim` and split it into `num_heads * head_dim`

**Comment by drisspg** [comment]
@huydhn I am sure you are very busy with the 2.1 release, but is it easy to determine if the CI/CD changes made on this PR are acceptable?

**Comment by huydhn** [comment]
> @huydhn I am sure you are very busy with the 2.1 release, but is it easy to determine if the CI/CD changes made on this PR are acceptable?

https://github.com/pytorch/pytorch/actions/runs/6005121633/job/16287206550 looks ok to me.  It took 42m to finish, which looks a bit higher than usual, but the build time depends a lot on caching and it fluctuates quite a bit https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=linux-focal-cuda12.1-py3.10-gcc9-sm86%20%2F%20build.  To unblock this, we could get this merged and monitor build time in trunk for sm86 jobs.

cc @clee2000 FYI

File: .circleci/scripts/binary_populate_env.sh [hunk 8]
```diff
@@ -156,7 +156,7 @@ EOL
 # nproc doesn't exist on darwin
 if [[ "$(uname)" != Darwin ]]; then
   # Because most Circle executors only have 20 CPUs, using more causes OOMs w/ Ninja and nvcc parallelization
-  MEMORY_LIMIT_MAX_JOBS=18
+  MEMORY_LIMIT_MAX_JOBS=12
```
Comment (huydhn) [inline .circleci/scripts/binary_populate_env.sh:159]: For posterity, please add a comment on why MAX_JOBS is set as such.  Otherwise, the context would be lost after a while :P

File: .ci/pytorch/build.sh (hunk 4 above)
Comment (huydhn) [reply .ci/pytorch/build.sh:165]: Please feel free to cut us an issue about this so we could look into this later to figure out if there is a better way to set it without slowing down the build.

**Review by huydhn — APPROVED** [review]
Approve the CI changed.

File: test/test_transformers.py [hunk 9]
```diff
@@ -1208,7 +1208,7 @@ class TestSDPAFailureModes(NNTestCase):
     @onlyCUDA
     @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not isSM86or89Device,
```
Comment (drisspg) [inline test/test_transformers.py:1209]: I have yet to update `PLATFORM_SUPPORTS_FUSED_SDPA` to exclude windows. I am surprised CI/CD is green, and think this indicates a lack of coverage for sm80 + on windows. But regardless I need audit uses of this, and maybe make a PLATFORM_SUPPORTS_FLASH_SDPA and PLATFORM_SUPPORTS_MEM_EFF_SDPA

File: test/test_transformers.py [hunk 10]
```diff
@@ -1473,6 +1473,50 @@ def test_mem_efficient_fail_bfloat16_sm50(self, device):
                 self.assertRaises(RuntimeError, lambda: torch.nn.functional.scaled_dot_product_attention(
                     q, k, v, None, 0.0, False))
 
+def _get_block_size(device, head_dim, is_causal):
```
Comment (drisspg) [inline test/test_transformers.py:1476]: taken from FlashAttention testing code

File: test/test_transformers.py [hunk 11]
```diff
@@ -1473,6 +1473,50 @@ def test_mem_efficient_fail_bfloat16_sm50(self, device):
...
+            return 128, 32
+    elif head_dim <= 192:
+        return (128, 64) if not is_dropout else (64, 64)
+    elif head_dim <= 224:
+        return (128, 64) if (is_sm80 or is_sm90) else (64, 64)
+    elif head_dim <= 256:
+        return (128, 64) if is_sm80 else (64, 64)
+
+
+def pad_last_dim(input_tensor, alignment_size, slice: bool = False):
```
Comment (drisspg) [inline test/test_transformers.py:1509]: port from c++

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 12]
```diff
@@ -40,54 +41,71 @@ std::tuple<Tensor, Tensor, Tensor> _flash_attention_backward(
...
-          is_causal,
-          num_splits,
-          philox_seed,
-          philox_offset
-  );
-  return std::make_tuple(dq, dk, dv);
+  // We check the whether the cumulative_sequence_length_q is defined
+  // in order to determine whether we are using varlen or dense forward
+  if (cumulative_sequence_length_q.defined()) {
+    // Varlen forward
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:67]: This path should be exercised by: https://github.com/pytorch/pytorch/pull/97485

File: .ci/pytorch/build.sh (hunk 4 above)
Comment (drisspg) [reply .ci/pytorch/build.sh:165]: https://github.com/pytorch/pytorch/issues/108174

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 13]
```diff
@@ -654,63 +655,43 @@ std::tuple<Tensor, Tensor, Tensor, Tensor, int64_t, int64_t, Tensor, Tensor, Ten
...
+              c10::nullopt,
+              c10::nullopt,
+              dropout_p,
+              is_causal,
+              return_debug_mask,
+              scale);
   // Reshape output to convert nnz to batch_size and seq_len
-  attention =
-      attention.view({batch_size, max_seqlen_batch_q, num_heads, head_dim}).transpose(1,2);
+  auto attention = output.transpose(1,2);
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:692]: bad use of auto

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 14]
```diff
@@ -31,6 +31,7 @@ namespace at {
 
 namespace native {
 
+// Expect inputs in format [ B, Seqlen, head_dims E]
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:34]: weird comment

File: aten/src/ATen/native/transformers/cuda/attention_backward.cu [hunk 15]
```diff
@@ -499,32 +517,20 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> _scaled_dot_product_flash_attenti
...
-  int64_t Nnz_kv{batch_size * max_seqlen_batch_k};
-
-  // For the standard MHA these will actually be views
-  Tensor query_reshaped = q_t.reshape({Nnz_q, num_heads, head_dim});
-  Tensor key_reshaped = k_t.reshape({Nnz_kv, num_heads, head_dim});
-  Tensor value_reshaped = v_t.reshape({Nnz_kv, num_heads, head_dim});
-
-  auto grad_out_reshaped = grad_out_.transpose(1,2).reshape({{Nnz_q, num_heads, head_dim}});
-  auto out_reshaped = out.transpose(1,2).reshape({Nnz_q, num_heads, head_dim});
+  auto grad_out_t = grad_out_.transpose(1,2);
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention_backward.cu:524]: more bad autos

**Review by drisspg — COMMENTED** [review]
okayyyy

File: .ci/pytorch/build.sh (hunk 4 above)
Comment (malfet) [reply .ci/pytorch/build.sh:165]: Err, should this be limited only to debug builds?

**Review by malfet — CHANGES_REQUESTED** [review]
It feels wrong to restrict all CUDA builds to just 2 threads, let's search for a ways to limit parallelism of a specific portion of the build, or reduce all builders to `linux.2xlarge` as right now majority of the builds are running on `linux.16xlarge`, which feels a huge waste of resources..

File: .ci/pytorch/build.sh (hunk 4 above)
Comment (drisspg) [reply .ci/pytorch/build.sh:165]: This is OOMing when building in with DEBUG=0

**Comment by huydhn** [comment]
> It feels wrong to restrict all CUDA builds to just 2 threads, let's search for a ways to limit parallelism of a specific portion of the build, or reduce all builders to `linux.2xlarge` as right now majority of the builds are running on `linux.16xlarge`, which feels a huge waste of resources..

This is CI build so I think it runs on `2xlarge`, so the decrease is from `nproc - 2 = 6` to `2` (https://github.com/pytorch/pytorch/blob/main/.github/workflows/_linux-build.yml#L144) for CUDA 8.6 build.  On the other hand, binary build uses 18 as set in https://github.com/pytorch/pytorch/blob/main/.circleci/scripts/binary_populate_env.sh#L158-L163, so the decrease there is from 18 to 12 I think as OOM also happens there (regular build, no debug)

**Comment by malfet** [comment]
> This is CI build so I think it runs on `2xlarge`, so the decrease is from `nproc - 2 = 6` to `2` (https://github.com/pytorch/pytorch/blob/main/.github/workflows/_linux-build.yml#L144) for CUDA 8.6 build. 

I'm ok with setting the limit with `proc // 3`, but I'm dead against hard-coding it to 2 or any other constant, because when somebody will try to speed up the build by moving it to a more powerful runner, they'll be confused as to why nothing happens. Also, can we please print a BIG WARNING before restricting number of threads?

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 16]
```diff
@@ -765,52 +746,102 @@ int64_t _fused_sdp_choice_cuda(const Tensor& query_, const Tensor& key, const Te
...
+      cumulative_sequence_length_q.has_value() ==
+          cumulative_sequence_length_k.has_value(),
+      "cumulative_sequence_length_q and cumulative_sequence_length_k must be both set or both not set");
+  TORCH_CHECK(
+      max_seqlen_batch_q.has_value() == max_seqlen_batch_k.has_value(),
+      "max_seqlen_batch_q and max_seqlen_batch_k must be both set or both not set");
+  Tensor output, q_padded, k_padded, v_padded, logsumexp, output_shape,
+      philox_seed, philox_offset, debug_attn_mask;
+  if (cumulative_sequence_length_q.has_value()) {
+    TORCH_CHECK(
```
Comment (cpuhrsch) [inline aten/src/ATen/native/transformers/cuda/attention.cu:780]: Do you want more checks than this?

**Review by cpuhrsch — APPROVED** [review]

**Comment by drisspg** [comment]
@malfet with the changes to how set the parallelism is this landable?  I think that I should probably not try and land this via ghfirst and instead import the diff, and wanted to make sure that this is okay before doing that

**Comment by drisspg** [comment]
I added the BC-Breaking label to this PR, since if users were running the below command on windows, prior to this PR this would pass and now this would not succeed. See: https://github.com/pytorch/pytorch/issues/108175
```
with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    torch.nn.functional.scaled_dot_product_attention(q,k,v)
```

**Comment by huydhn** [comment]
@pytorchbot revert -m 'I think we miss a case here for sm80 build on inductor workflow as it is now OOM on trunk https://github.com/pytorch/pytorch/actions/runs/6042843139' -c nosignal

Here is another example https://github.com/pytorch/pytorch/actions/runs/6043200135/job/16399783493 where Docker exited with exit code 137 (just another more graceful way to say it's OOM).  I will add `ciflow/inductor` to trigger inductor jobs on the PR

**Comment by huydhn** [comment]
There is also this failure on Windows GPU which seems to be related to this change https://hud.pytorch.org/pytorch/pytorch/commit/06d74e6b24a6f7dfd0b8d05e2171d74def166140

```
_______ TestSDPAFailureModesCUDA.test_flash_autocast_fp32_bfloat16_cuda _______
Traceback (most recent call last):
  File "C:\actions-runner\_work\pytorch\pytorch\test\test_transformers.py", line 1409, in test_flash_autocast_fp32_bfloat16
    _ = torch.nn.functional.scaled_dot_product_attention(
RuntimeError: No available kernel.  Aborting execution.

To execute this test, run the following from the base repo dir:
     python test\test_transformers.py -k test_flash_autocast_fp32_bfloat16_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
---------------------------- Captured stderr call -----------------------------
C:\actions-runner\_work\pytorch\pytorch\test\test_transformers.py:1409: UserWarning: 1Torch was not compiled with flash attention. (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\aten\src\ATen\native\transformers\cuda\sdp_utils.cpp:234.)
  _ = torch.nn.functional.scaled_dot_product_attention(
C:\actions-runner\_work\pytorch\pytorch\test\test_transformers.py:1409: UserWarning: Memory efficient kernel not used because: (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\aten\src\ATen\native\transformers\cuda\sdp_utils.cpp:350.)
  _ = torch.nn.functional.scaled_dot_product_attention(
C:\actions-runner\_work\pytorch\pytorch\test\test_transformers.py:1409: UserWarning: Memory Efficient attention has been runtime disabled. (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\aten\src\ATen/native/transformers/sdp_utils_cpp.h:437.)
  _ = torch.nn.functional.scaled_dot_product_attention(
C:\actions-runner\_work\pytorch\pytorch\test\test_transformers.py:1409: UserWarning: Flash attention kernel not used because: (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\aten\src\ATen\native\transformers\cuda\sdp_utils.cpp:352.)
  _ = torch.nn.functional.scaled_dot_product_attention(
C:\actions-runner\_work\pytorch\pytorch\test\test_transformers.py:1409: UserWarning: 0Torch was not compiled with flash attention. (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\aten\src\ATen\native\transformers\cuda\sdp_utils.cpp:234.)
  _ = torch.nn.functional.scaled_dot_product_attention(
- generated xml file: C:\actions-runner\_work\pytorch\pytorch\test\test-reports\python-pytest\test_transformers\test_transformers-ddbe1b291998f64d.xml -
=========================== short test summary info ===========================
FAILED [0.0016s] test_transformers.py::TestSDPAFailureModesCUDA::test_flash_autocast_fp32_bfloat16_cuda
```

I remember that flash attention isn't available on Windows.

**Comment by drisspg** [comment]
@huydhn  The windows is real, and I know what change to make for that
I think I also need know to check for 8.0 and 8.6 in the build.sh ?

**Comment by huydhn** [comment]
> @huydhn The windows is real, and I know what change to make for that I think I also need know to check for 8.0 and 8.6 in the build.sh ?

Yeah, I think so, needs to apply the same MAX_JOBS hack for both.

**Comment by huydhn** [comment]
I need to massage the revert commit https://github.com/pytorch/pytorch/pull/108827 a bit to keep trunk green

**Comment by drisspg** [comment]
@huydhn so you think if I re-emerge through ghfirst/pytorchbot or do I need to import the diff?

**Comment by huydhn** [comment]
> @huydhn so you think if I re-emerge through ghfirst/pytorchbot or do I need to import the diff?

There is nothing wrong with importing the diff now.  If we want, we could land it internally then OSS will be landed.  Otherwise, the PR can be merged normally and this will go through ghfirst diff train.

**Comment by drisspg** [comment]
@huydhn It wasn't letting me import the DIff so I unlinked and re-to D49137501

**Comment by fxmarty** [comment]
Hi, this probably won't land in PyTorch 2.1 @drisspg?

**Comment by drisspg** [comment]
@fxmarty Yeah unfortunately this is not going to land in 2.1 and will available in nightlies. There was some internal changes needed that weren't synced properly, causing the revert which I am hoping to have remedied this morning.

**Comment by vince62s** [comment]
@drisspg will this be part of pt2.1.1 or 2.2 ?

**Comment by drisspg** [comment]
@vince62s since 2.1.1 will be reserved for bug fixes this is unlikely to be included there and this will be added in 2.2.

**Comment by vince62s** [comment]
is there a already a release date for 2.2 ?
this is unfortunate, because flash2 was integrated in a timely manner and we have to wait ...:)

**Comment by drisspg** [comment]
> is there a already a release date for 2.2 ? this is unfortunate, because flash2 was integrated in a timely manner and we have to wait ...:)

There is are recent dev-discuss post on the release cadence here:
https://dev-discuss.pytorch.org/t/release-cadence-for-year-2023-2024/1557

**Comment by daixiangzi** [comment]
can you offer test_time flash v1 and v2 code?

**Comment by michaelshekasta** [comment]
@drisspg 
Is FA2 SDPA stable in the latest PyTorch version? Also, I recall FA2 boasting a fourfold increase in runtime efficiency. However, I recently came across information suggesting it only provides a twofold improvement. How do these figures align? (Source: https://github.com/Dao-AILab/flash-attention)

**Comment by drisspg** [comment]
I'm not sure what you mean by stable, but yes FAv2 is one of the backends for sdpa in the stable release. It is currently the highest priority and when all conditions are met it will be run. 

The increase from Fav1 to fav2 is ~ 2x 

I am not sure where the 4x number is that you are seeing

**Comment by michaelshekasta** [comment]
@drisspg  , thank you for your response!! very appreciated it !

1 - By "stable," I mean that the feature is now officially part of the PyTorch stable release. So, if I understand correctly, SDPA FA2 is now stable in the latest version, correct?

2 - When comparing "native PyTorch" to "FA2," I see that the latter is four times as efficient.

**Comment by drisspg** [comment]
1, yup FaV2 is part of the latest release
2. The comparison is against riding the equivalent implementation in more primitive PyTorch ops (our math backend)

**Comment by michaelshekasta** [comment]
@drisspg Thanks again!
1 - You are so awesome! 
2 - I am talking about this comparison (blue bars vs. purple bars):
image

**Comment by drisspg** [comment]
There are some benchmark scripts comparing the perf of different backends here: https://github.com/pytorch/pytorch/blob/91bf952d10e9524a9b078900d9807efa5d252f5c/benchmarks/transformer/sdpa.py#L129

tldr is that the chart above is comparing SDPBackend.Math vs SDPBackend.FLASH_ATTENTION

there is some more docs here: https://pytorch.org/docs/stable/generated/torch.nn.attention.sdpa_kernel.html#torch.nn.attention.sdpa_kernel

**Comment by orena1** [comment]
> There are some benchmark scripts comparing the perf of different backends here:
> 
> https://github.com/pytorch/pytorch/blob/91bf952d10e9524a9b078900d9807efa5d252f5c/benchmarks/transformer/sdpa.py#L129
> 
> tldr is that the chart above is comparing SDPBackend.Math vs SDPBackend.FLASH_ATTENTION
> 
> there is some more docs here: https://pytorch.org/docs/stable/generated/torch.nn.attention.sdpa_kernel.html#torch.nn.attention.sdpa_kernel

Thanks but how did you compare falshattention v1 to flashattention v2?

I only see those:
```
ERROR: An error occurred when trying to determine the backend.

MATH: The math backend for scaled dot product attention.

FLASH_ATTENTION: The flash attention backend for scaled dot product attention.

EFFICIENT_ATTENTION: The efficient attention backend for scaled dot product attention.

CUDNN_ATTENTION: The cuDNN backend for scaled dot product attention.
```
